masa/utils/uids.py

```diff
@@ -18,57 +21,116 @@ def check_uid_availability(
+            if response.dendrite.status_code == 200
+        ]
+        failed_uids = [
+            uid
+            for uid, response in zip(uids, responses)
+            if response.dendrite.status_code != 200
+        ]
+    except Exception as e:
+        bt.logging.error(f"Dendrite ping failed: {e}")
+        successful_uids = []
+        failed_uids = uids
+    bt.logging.debug(f"ping() successful uids: {successful_uids}")
+    bt.logging.debug(f"ping() failed uids    : {failed_uids}")
+    return successful_uids, failed_uids
+
+def filter_duplicated_axon_ips_for_uids(uids, metagraph):
+    ips = []
+    miner_ip_filtered_uids = []
+    for uid in uids:
+        if metagraph.axons[uid].ip not in ips:
+            ips.append(metagraph.axons[uid].ip)
+            miner_ip_filtered_uids.append(uid)
+    return miner_ip_filtered_uids
+
+async def get_random_uids(
     self, k: int, exclude: List[int] = None
 ) -> torch.LongTensor:
-    """Returns k available random uids from the metagraph.
+    """
+    Returns at most k available random uids from the metagraph.
+
     Args:
         k (int): Number of uids to return.
         exclude (List[int]): List of uids to exclude from the random sampling.
     Returns:
         uids (torch.LongTensor): Randomly sampled available uids.
     Notes:
-        If `k` is larger than the number of available `uids`, set `k` to the number of available `uids`.
+        If `k` is larger than the number of available `uids`, set `k` to the number of
+        available `uids`.
     """
-    candidate_uids = []
-    avail_uids = []
+    dendrite = bt.dendrite(wallet=bt.wallet)
     
     print("get random uids")
 
-    for uid in range(self.metagraph.n.item()):
-        uid_is_available = check_uid_availability(
-            self.metagraph, uid, self.config.neuron.vpermit_tao_limit
-        )
-        uid_is_not_excluded = exclude is None or uid not in exclude
-
-        if uid_is_available:
-            avail_uids.append(uid)
-            if uid_is_not_excluded:
-                candidate_uids.append(uid)
-    # If k is larger than the number of available uids, set k to the number of available uids.
-    k = min(k, len(avail_uids))
-    # Check if candidate_uids contain enough for querying, if not grab all avaliable uids
-    available_uids = candidate_uids
-    
-    
-    print("AVAIL UIDS: ", avail_uids)
-    if len(candidate_uids) < k:
-        available_uids += random.sample(
-            [uid for uid in avail_uids if uid not in candidate_uids],
-            k - len(candidate_uids),
-        )
+    try:
+        # Generic sanitation
+        avail_uids = get_available_uids(self.metagraph, self.config.neuron.vpermit_tao_limit)
+        candidate_uids = remove_excluded_uids(avail_uids, exclude)
+
+        healthy_uids, _ = await ping_uids(dendrite, self.metagraph, candidate_uids)
+        filtered_uids = filter_duplicated_axon_ips_for_uids(healthy_uids, self.metagraph)
+
```

Comment:
I feel like this can be stored in the validator's instance and recalculated from time to time instead of being calculated on every call

wdyt @juanmanso?

Reply:
Yep, I think the logic can be a bit more thought to optimize, otherwise saving the healthy ones would make that the unhealthy be somewhat blacklisted.

I'd suggest to think about some cooldown + blacklisting functions to be able to decrease the amount of pings (which I agree it's an improvement).

If you agree @hide-on-bush-x we can close this and you can follow-up, how does that sound?